Week10/app/src/main/java/com/em4n0101/mytvshows/app/ViewModelsModule.kt

```diff
@@ -0,0 +1,78 @@
+package com.em4n0101.mytvshows.app
+
+import com.em4n0101.mytvshows.view.cast.CastMemberActivity
+import com.em4n0101.mytvshows.view.episodes.EpisodesActivity
+import com.em4n0101.mytvshows.view.schedule.ScheduleFragment
+import com.em4n0101.mytvshows.view.searchshow.SearchShowFragment
+import com.em4n0101.mytvshows.view.showdetail.ShowDetailActivity
+import com.em4n0101.mytvshows.view.usercastmembers.UserFavoriteCastMemberFragment
+import com.em4n0101.mytvshows.view.usershows.UserFavoriteShowsFragment
+import com.em4n0101.mytvshows.viewmodel.cast.CastMemberViewModel
+import com.em4n0101.mytvshows.viewmodel.episodes.EpisodesViewModel
+import com.em4n0101.mytvshows.viewmodel.schedule.ScheduleViewModel
+import com.em4n0101.mytvshows.viewmodel.searchshow.SearchShowViewModel
+import com.em4n0101.mytvshows.viewmodel.showdetails.ShowDetailsViewModel
+import com.em4n0101.mytvshows.viewmodel.usercastmember.UserFavoriteCastMemberViewModel
+import com.em4n0101.mytvshows.viewmodel.usershows.UserFavoriteShowsViewModel
+import org.koin.android.viewmodel.dsl.viewModel
+import org.koin.core.qualifier.named
+import org.koin.dsl.module
+
+const val SCOPE_SEARCH_SHOW = "SCOPE_SEARCH_SHOW"
+const val SCOPE_SHOW_DETAILS = "SCOPE_SHOW_DETAILS"
+const val SCOPE_EPISODES = "SCOPE_EPISODES"
+const val SCOPE_CAST_MEMBERS = "SCOPE_CAST_MEMBERS"
+const val SCOPE_SCHEDULE = "SCOPE_SCHEDULE"
+const val SCOPE__FAVORITE_CAST_MEMBERS = "SCOPE__FAVORITE_CAST_MEMBERS"
+const val SCOPE_FAVORITE_SHOWS = "SCOPE_FAVORITE_SHOWS"
+
+var viewModelsModule = module {
+    scope(named(SCOPE_SEARCH_SHOW)) {
+         scoped {
+            SearchShowFragment()
+         }
+         viewModel { SearchShowViewModel(get()) }
+    }
+
+    scope(named(SCOPE_SHOW_DETAILS)) {
+        scoped {
+            ShowDetailActivity()
+        }
+        viewModel { ShowDetailsViewModel(get()) }
+    }
+
+    scope(named(SCOPE_EPISODES)) {
+        scoped {
+            EpisodesActivity()
+        }
+        viewModel { EpisodesViewModel(get()) }
+    }
+
+    scope(named(SCOPE_CAST_MEMBERS)) {
+        scoped {
+            CastMemberActivity()
+        }
+        viewModel { CastMemberViewModel(get()) }
+    }
+
+    scope(named(SCOPE_SCHEDULE)) {
+        scoped {
+            ScheduleFragment()
+        }
+        viewModel { ScheduleViewModel(get()) }
+    }
+
+    scope(named(SCOPE__FAVORITE_CAST_MEMBERS)) {
+        scoped {
+            UserFavoriteCastMemberFragment()
+        }
+        viewModel { UserFavoriteCastMemberViewModel(get()) }
+    }
+
+    scope(named(SCOPE_FAVORITE_SHOWS)) {
+        scoped {
+            UserFavoriteShowsFragment()
+        }
+        viewModel { UserFavoriteShowsViewModel(get()) }
+    }
+}
```

Comment:
I've found some fundamental issues, which I think are here because it seems you are not that clear on what the concept of scopes is about, and how to use them.

The scope in this case is like a time-limited container in which instances are being held. Let's take an example with the login feature like in the homework document. 
You can create a custom scope that is going to hold and provide dependencies as long as `LoginActivity` is alive. You do that by creating a custom scope in LoginActivity `onCreate()` method and closing the scope in `onDestroy()` method. Or, you can use `lifecycleScope` extension property on `LifecycleOwner`.

The other thing that is wrong here is that for some reason, you're trying to create instances of Activity and Fragment classes. We can't do that, because the system is responsible for instantiating those. Think about the way you start activities. You create an Intent and then call `startActivity(intent)`, you don't create Activity instance manually. 

One other thing to note here is that ViewModel classes are by default scoped to the lifecycle owner in whom they are injected. So, if you inject ViewModel into an Activity, it will stay alive until the Activity is destroyed (not when config changes occur). You don't have to create scopes just for ViewModels (but you can if you want, and I show this in the homework solution)

With the creation of so many Activity and Fragment instances, you are creating a huge memory overhead. These are expensive objects. 

To conclude this: You don't create custom scopes to scope activities and fragments, you create scopes to scope dependencies for activities and fragments. Activities and Fragments are there to determine when the scope is going to be created and destroyed.

Reply:
You can make the module as `val` instead of `var`